src/java.base/share/classes/java/util/stream/AbstractPipeline.java

```diff
@@ -466,7 +466,32 @@ final StreamShape getSourceShape() {
 
     @Override
     final <P_IN> long exactOutputSizeIfKnown(Spliterator<P_IN> spliterator) {
-        return StreamOpFlag.SIZED.isKnown(getStreamAndOpFlags()) ? spliterator.getExactSizeIfKnown() : -1;
+        int flags = getStreamAndOpFlags();
+        long size = StreamOpFlag.SIZED.isKnown(flags) ? spliterator.getExactSizeIfKnown() : -1;
+        // Currently, we have no stateless SIZE_ADJUSTING intermediate operations,
+        // so we can simply ignore SIZE_ADJUSTING in parallel more, as adjustments
+        // are already accounted in the input spliterator.
+        //
+        // If we ever have SIZE_ADJUSTING intermediate operation,
+        // we would need step back until depth == 0, then call exactOutputSize() for
+        // the subsequent stages.
```

Comment:
```suggestion
        // Currently, we have no stateless SIZE_ADJUSTING intermediate operations,
        // so we can simply ignore SIZE_ADJUSTING in parallel streams, since adjustments
        // are already accounted in the input spliterator.
        //
        // If we ever have a stateless SIZE_ADJUSTING intermediate operation,
        // we would need step back until depth == 0, then call exactOutputSize() for
        // the subsequent stages.
```